cp1616/include/cp1616/cp1616_io_controller.h

```diff
@@ -0,0 +1,278 @@
+   *
+   * \return error_code if succeded (see pnioerrx.h for detailed description)
+   */
+          int init();
+
+  /**
+   * \brief Closes and uninitializes IO Controller
+   *
+   * \return error_code if succeded (see pnioerrx.h for detailed description)
+   */
+          int uinit();
+
+   /**
+    * \brief Initialize Controller data structures
+    */
+          void configureControllerData();
+
+   /**
+   * \brief Adds input module to Controller data structre
+   *
+   * \param input_size  input data length according to STEP7 setup
+   * \param input_start_address I address according to STEP7 setup
+   *
+   * \return 0 if input module added successfully
+   */
+          int addInputModule(unsigned int input_size, unsigned int input_start_address);
+  /**
+   * \brief Adds output module to Controller data structure
+   *
+   * \param output_size output data length according to STEP7 setup
+   * \param output_start_address Q address according to STEP7 setup
+   *
+   * \return 0 if input module added successfully
+   */
+
+          int addOutputModule(unsigned int output_size, unsigned int output_start_address);
+
+  /**
+   * \brief Reads recieved data from IO Base layer
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int updateCyclicInputData();
+
+  /**
+   * \brief Writes data to IO Base library layer for transmission
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int updateCyclicOutputData();
+
+  /**
+   * \brief Changes IO Controller mode
+   *
+   * \param requested_mode OFFLINE/OPERATE/CLEAR
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int changeAndWaitForPnioMode(PNIO_MODE_TYPE requested_mode);
+
+  /**
+   * \brief Sets IO Controller OutData
+   *
+   * \param module output module index according to addOutputModule order
+   * \param data_index index of out_data_
+   * \param value requested value
+   */
+          void setOutData(unsigned int module, unsigned int data_index, PNIO_UINT8 value);
+
+  /**
+   * \brief Sets IO Controller cp_ready_ variable
+   *
+   * \param cp_ready_value
+   */
+          void setCpReady(int cp_ready_value);
+
+  /**
+   * \brief Returns IO Controller cp_ready_ state (used in callbackForAlarmIndication)
+   *
+   * \return cp_ready_ flag
+   */
+          int getCpReady();
+
+  /**
+   * \brief Sets IO Controller cp_current_mode_ flag (used in callbackForModeChangeIndication)
+   *
+   * \param mode - current mode
+   */
+          void setCpCurrentModeFlag(PNIO_MODE_TYPE mode);
+
+  /**
+   * \brief Returns IO Controller cp_current_mode_ flag
+   *
+   * \return cp_current_mode_
+   */
+          PNIO_MODE_TYPE getCpCurrentModeFlag();
+
+  /**
+   * \brief Sets sem_mode_change_ flag (used in callbackForModeChangeIndication)
+   *
+   * \param value
+   */
+          void setSemModChange(int mod_change);
+
+  /** Debugging functions  */
```

Comment:
I would recommend keeping this around...not just for debugging.  This will be useful for error reporting and unit testing.